File: plugins/cbre-il/skills/cbre-property-longlist/evals/f26_gla_single_derivation_test.mjs
// evals/f26_gla_single_derivation_test.mjs - execute the REAL chrome from a BUILT dashboard and
// prove that every surface printing the "Total GLA" label prints the SAME figure. (F26)
// Offline; no npm deps.  Usage: node f26_gla_single_derivation_test.mjs <built_html_path>
import fs from 'node:fs';
import vm from 'node:vm';

const htmlPath = process.argv[2];
if (!htmlPath) { console.error('usage: node f26_gla_single_derivation_test.mjs <built_html>'); process.exit(2); }
const html = fs.readFileSync(htmlPath, 'utf8');

const scripts = [...html.matchAll(/<script(?![^>]*\bsrc=)[^>]*>([\s\S]*?)<\/script>/gi)].map(m => m[1]);
const NAMES = ['PROPS', 'glaVal', 'glaStr', 'glaUnit', 'statedTotal', 'statedTotalHTML', 'cardHTML',
               'detailHTML', 'compareHTML', 'totalAnnualRent', 'AREA_UNIT', 'T'];
const code = scripts.join('\n;\n') + '\n;\n' +
  NAMES.map(n => `__capture__(${JSON.stringify(n)}, typeof ${n} !== "undefined" ? ${n} : undefined);`).join('\n') + '\n';

// See modal_render_test.mjs for why this sandbox is shaped the way it is.
const sink = new Proxy(function () {}, {
  get: (_t, p) => (p === Symbol.toPrimitive || p === 'toString' || p === 'valueOf') ? () => '' : sink,
  apply: () => sink, construct: () => sink, has: () => true,
});
const target = { console };
for (const name of Object.getOwnPropertyNames(globalThis)) {
  if (!(name in target)) { try { target[name] = globalThis[name]; } catch { /* ignore */ } }
}
target.globalThis = target;
target.__capture__ = (name, val) => { target[name] = val; };
const ctx = vm.createContext(new Proxy(target, {
  get: (t, p) => (p in t ? t[p] : sink),
  has: () => true,
}));
try { vm.runInContext(code, ctx, { filename: 'built.inline.js' }); }
catch (e) { console.error('FAIL: template script threw during eval:', e && e.message); process.exit(1); }

const { PROPS: props, glaVal, glaStr, glaUnit, statedTotal, statedTotalHTML, cardHTML, detailHTML,
        compareHTML, totalAnnualRent, AREA_UNIT, T } = target;
for (const [n, f] of Object.entries({ glaVal, glaStr, glaUnit, statedTotal, statedTotalHTML, cardHTML, detailHTML, T })) {
  if (typeof f !== 'function') { console.error(`FAIL: could not capture ${n}() from the built chrome`); process.exit(1); }
}
if (!Array.isArray(props)) { console.error('FAIL: PROPS not captured'); process.exit(1); }

const fails = [];
const ck = (ok, label) => { console.log((ok ? '  ok   ' : '  FAIL ') + label); if (!ok) fails.push(label); };
const esc = s => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
const LABEL = T('row_total_gla');
// the modal / compare cell that carries the Total GLA label -> its printed value
const rowValue = h => {
  const m = h.match(new RegExp(`<div class="spec-k">${esc(LABEL)}</div><div class="spec-v">([^<]*)</div>`));
  return m ? m[1] : null;
};
const cardSub = p => (cardHTML(p).match(/<small class="spec-total">([^<]*)</) || [])[1] || null;

const disagrees = props.find(p => p.park === 'Disagrees');
const agrees = props.find(p => p.park === 'Agrees');
const unstated = props.find(p => p.park === 'Unstated');
if (!disagrees || !agrees || !unstated) { console.error('FAIL: fixture properties missing'); process.exit(1); }

console.log('-- ONE figure under the one label, on a property whose source printed a larger total --');
const modalD = detailHTML(disagrees);
ck(glaVal(disagrees) === 45649, `glaVal() is the source's printed total (${glaVal(disagrees)})`);
ck(glaUnit(disagrees) === 'sq ft', `glaUnit() is the stated unit (${glaUnit(disagrees)})`);
ck(rowValue(modalD) === `45,649 sq ft`, `the MODAL row "${LABEL}" prints ${JSON.stringify(rowValue(modalD))}`);
ck(cardSub(disagrees) === `${LABEL} 45,649 sq ft`, `the CARD qualifier prints ${JSON.stringify(cardSub(disagrees))}`);
ck(rowValue(modalD) === glaStr(disagrees), 'the modal row IS glaStr(): no second derivation');
ck(!/42,000/.test(modalD), 'the summed 42,000 (warehouse + office) appears NOWHERE in the modal');
ck(!/42,000/.test(cardHTML(disagrees)), '...nor on the card');
if (typeof compareHTML === 'function') {
  let cmp = null;
  try { cmp = compareHTML([disagrees, agrees]); } catch (e) { cmp = null; }
  if (typeof cmp === 'string' && cmp.length) {
    ck(/45,649 sq ft/.test(cmp) && !/42,000/.test(cmp), 'the COMPARE matrix prints the stated total, never the sum');
  } else {
    console.log('  skip compareHTML needs a DOM here; its row is pinned structurally by the Python driver');
  }
}

console.log('\n-- the fallback is untouched where no stated total was surfaced --');
ck(glaVal(agrees) === 30000, `agreeing property (stated inside the tolerance, nothing attached): ${glaVal(agrees)}`);
ck(rowValue(detailHTML(agrees)) === `30,000 ${AREA_UNIT}` && cardSub(agrees) === null,
   'modal prints the sum, card prints no qualifier');
ck(glaVal(unstated) === 20000 && rowValue(detailHTML(unstated)) === `20,000 ${AREA_UNIT}`,
   'unstated total: warehouse + office, in the dataset unit');
ck(statedTotal({}) === null && statedTotal({ preBaked: { statedTotal: { value: 'tbd' } } }) === null
   && statedTotal({ preBaked: { statedTotal: { value: 0 } } }) === null,
   'statedTotal() refuses a missing, sentinel or non-positive figure');
ck(glaVal({ warehouseArea: 'tbd' }) === null && glaStr({ warehouseArea: 'tbd' }) === '\u2014',
   'no numeric warehouse area and no stated total -> null / the dash, exactly as before');
ck(glaVal({ warehouseArea: 1000, officeAreaVal: -5 }) === 1000 && glaVal({ warehouseArea: 1000, officeAreaVal: 250 }) === 1250,
   'the fallback keeps the > 0 office guard the arithmetic gate replicates');

console.log('\n-- money is NOT re-derived from the stated total --');
if (typeof totalAnnualRent === 'function') {
  ck(totalAnnualRent(disagrees) === 42000 * 9.5,
     `totalAnnualRent still multiplies the SUMMED areas (${totalAnnualRent(disagrees)}), never a total the source did not price`);
}

console.log('STATUS:', fails.length ? 'BLOCKED' : 'ALL-PASS');
process.exit(fails.length ? 1 : 0);

File: plugins/cbre-il/skills/cbre-property-longlist/evals/f26_gla_single_derivation_test.py
#!/usr/bin/env python3
"""f26_gla_single_derivation_test.py - one "Total GLA" label, ONE derivation, on every surface. (F26)

WHAT WAS WRONG. The card printed the source's own stated total (preBaked.statedTotal) as a
qualifier, while the detail modal and the compare matrix computed warehouse + office under the
SAME "Total GLA" label. Wherever a deck's total includes space the two summed fields do not hold
(a gatehouse, a hub office, a plant room) the two disagreed: on a live run, four of nine
properties, by 16,845 / 10,692 / 5,000 / 4,269 sq ft. A blind reviewer confirmed against the
bound site plans' schedules of accommodation that the CARD figure (the stated total) is right.

THE FIX, and what it is not. glaVal() adopts the stated total when the builder surfaced one and
falls back to warehouse + office otherwise; glaUnit() carries the matching unit; the card, the
modal row and the compare cell all read those two. NO DATA CHANGED: an earlier attempt to make
the numbers agree by inflating an office area was refused because it would misstate a printed
figure, and totalAnnualRent() still multiplies the summed areas (a rate applied to a total the
source did not price would print a rent no source stated).

Offline. Needs node to execute the chrome (as statedtotal_card_test does).
"""
from __future__ import annotations

import hashlib
import json
import re
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
HELPERS = ROOT / "helpers"
sys.path.insert(0, str(HELPERS))
import _common as C  # noqa: E402
import build_dashboard as BD  # noqa: E402

PX = ("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0"
      "lEQVR42mNkYPhfDwAChwGA60e6kgAAAABJRU5ErkJggg==")
FAILS: list[str] = []


def ck(ok, msg):
    print(("  [PASS] " if ok else "  [FAIL] ") + msg)
    if not ok:
        FAILS.append(msg)


def _canon():
    base = {"country": "GB", "developer": "Dev", "city": "Sometown", "status": "Available",
            "photo": PX, "gallery": [PX], "lat": 52.49, "lng": -0.69, "areaUnit": "sq ft",
            "rentUnit": "\u00a3/sq ft/yr", "warehouseRent": "\u00a39.50 / sq ft / year",
            "warehouseRentVal": 9.5}
    props = [
        # 40,000 + 2,000 = 42,000 summed vs 45,649 printed -> the stated total is surfaced
        dict(id=1, park="Disagrees", warehouseArea=40000, officeArea="2000", officeAreaVal=2000),
        # 30,000 summed vs 30,010 printed -> inside the gate's tolerance, nothing surfaced
        dict(id=2, park="Agrees", warehouseArea=30000),
        # no stated total at all
        dict(id=3, park="Unstated", warehouseArea=20000),
    ]
    stated = {"1": {"value": 45649, "unit": "sq ft", "source_file": "t.xlsx", "locator": "S!r2"},
              "2": {"value": 30010, "unit": "sq ft", "source_file": "t.xlsx", "locator": "S!r3"}}
    return {"meta": {"client": "SingleDerivation",
                     "units": {"area": "sq ft", "rent": "\u00a3/sq ft/yr"},
                     "hero": {"topbar_meta": "", "eyebrow": "", "title_html": "",
                              "lede": "", "footer_copyright": ""},
                     "statedTotals": stated},
            "pois": [], "regions": {},
            "properties": [dict(base, **p) for p in props]}


def main() -> int:
    tpl = C.load_template()

    print("== structure: one reader of the stated total, and three surfaces reading glaVal/glaUnit ==")
    ck(tpl.count("p.preBaked.statedTotal") == 1,
       f"exactly ONE expression in the chrome reads p.preBaked.statedTotal ({tpl.count('p.preBaked.statedTotal')})")
    ck("function statedTotal(p){" in tpl and "function glaUnit(p){" in tpl,
       "statedTotal() and glaUnit() exist")
    gv = tpl.split("function glaVal(p){", 1)[-1][:400]
    ck("const st = statedTotal(p);" in gv and "if(st) return st.value;" in gv,
       "glaVal() adopts the stated total first...")
    ck("p.officeAreaVal" in gv and "return w + o;" in gv,
       "...and falls back to warehouse + office (the arithmetic gate's derivation)")
    # v45: the null branch prints BLANK, the chrome's one absence token, where it used to
    # print a long dash of its own. The derivation and the unit are byte-identical to v41.
    ck("function glaStr(p){ const g = glaVal(p); return g == null ? BLANK : fmt(g) + \" \" + glaUnit(p); }" in tpl,
       "glaStr() prints glaVal() in glaUnit(), and BLANK when there is neither")
    sth = tpl.split("function statedTotalHTML(p){", 1)[-1][:300]
    ck("${fmt(glaVal(p))} ${glaUnit(p)}" in sth,
       "the CARD qualifier prints glaVal()/glaUnit(): the same call as the modal row")
    ck("row(T('row_total_gla'), glaStr(p))," in tpl, "the MODAL row prints glaStr()")
    ck("[T('row_total_gla'), p=>glaStr(p), 'warehouseArea']," in tpl, "the COMPARE cell prints glaStr()")
    n_label = len(re.findall(r"""T\(["']row_total_gla["']\)""", tpl))
    ck(n_label == 3,
       f"the label is RENDERED on exactly three surfaces (card qualifier, modal, compare); found {n_label}")
    tar = tpl.split("function totalAnnualRent(p){", 1)[-1][:700]
    ck("glaVal(" not in tar and "(wa + oa) * wr" in tar,
       "totalAnnualRent() does NOT read glaVal(): money is never re-derived from a stated total")

    print()
    print("== the builder's attach still compares against the FALLBACK derivation ==")
    props = [C.fill_render_sentinels(dict(p)) for p in _canon()["properties"]]
    BD._attach_stated_totals(props, _canon()["meta"])
    got = {p["id"]: (p.get("preBaked") or {}).get("statedTotal") for p in props}
    ck(got[1] == {"value": 45649, "unit": "sq ft"}, f"the disagreeing property carries the stated total {got[1]}")
    ck(got[2] is None and got[3] is None, "inside-tolerance and unstated carry nothing")

    print()
    print("== the version bump and the hash chain the template contract requires ==")
    label = (ROOT / "assets" / "VERSION").read_text(encoding="utf-8").splitlines()[0].strip()
    ck(int(re.sub(r"\D", "", label) or 0) >= 41, f"assets/VERSION is at/after v41 ({label})")
    expected = hashlib.sha256(tpl.encode("utf-8")).hexdigest()
    recorded = C.load_version().get("chrome_sha256", "")
    ck(recorded == expected,
       f"VERSION's chrome_sha256 is the live template's text hash (recorded {recorded[:12]}, computed {expected[:12]})")

    print()
    print("== a real build, executed ==")
    with tempfile.TemporaryDirectory() as td:
        d = Path(td)
        cp, hp = d / "c.json", d / "b.html"
        cp.write_text(json.dumps(_canon()), encoding="utf-8")
        BD.build(cp, hp)
        built = hp.read_text(encoding="utf-8")
        again, _ = BD.render(C.load_canonical(cp))
        ck(again == built, "re-rendering the same canonical is byte-identical (validate-html's floor)")
        node = shutil.which("node") or r"C:\Users\TBaaij\nodejs\node.exe"
        mjs = Path(__file__).with_suffix(".mjs")
        if not Path(node).exists() and not shutil.which("node"):
            ck(False, "node is required to execute the chrome (install node or add it to PATH)")
        else:
            r = subprocess.run([node, str(mjs), str(hp)], capture_output=True, text=True,
                               encoding="utf-8", errors="replace")
            print((r.stdout or "").rstrip())
            if r.returncode != 0:
                print((r.stderr or "").rstrip())
            ck(r.returncode == 0, "the EXECUTED cross-surface assertions pass (see above)")

    print()
    if FAILS:
        print(f"F26 GLA SINGLE DERIVATION TEST: FAIL ({len(FAILS)})")
        return 1
    print("F26 GLA SINGLE DERIVATION TEST: PASS")
    return 0


if __name__ == "__main__":
    sys.exit(main())
